packages/flutter/lib/src/widgets/shortcuts.dart

```diff
@@ -1197,3 +1199,73 @@ class _ShortcutsMarker extends InheritedNotifier<ShortcutManager> {
 
   ShortcutManager get manager => super.notifier!;
 }
+
+/// A widget that provides an uncomplicated mechanism for binding a key
+/// combination to a specific callback.
+///
+/// This is similar to the functionality provided by the [Shortcuts] widget, but
+/// instead of requiring a mapping to an [Intent], and an [Actions] widget
+/// somewhere in the widget tree to bind the [Intent] to, it just takes a set of
+/// bindings that bind the key combination directly to a [VoidCallback].
+///
+/// Because it is a simpler mechanism, it doesn't provide the ability to disable
+/// the callbacks, or to separate the definition of the shortcuts from the
+/// definition of the code that is triggered by them (the role that actions play
+/// in the [Shortcuts]/[Actions] system).
+///
+/// However, for some applications the complexity and flexibility of the
+/// [Shortcuts] and [Actions] mechanism is overkill, and this widget is here for
+/// those apps.
+///
+/// [Shortcuts] and [CallbackShortcuts] can both be used in the same app.
+class CallbackShortcuts extends StatelessWidget {
+  /// Creates a const [CallbackShortcuts] widget.
+  const CallbackShortcuts({
+    Key? key,
+    required this.bindings,
+    required this.child,
+  }) : super(key: key);
+
+  /// A map of key combinations to callbacks used to define the shortcut
+  /// bindings.
+  ///
+  /// If a descendant of this widget has focus, and a key is pressed, the
+  /// activator keys of this map will be asked if they accept the key event. If
+  /// they do, then the corresponding callback is invoked, and the key event
+  /// propagation is halted. If none of the activators accept the key event,
+  /// then the key event continues to be propagated up the focus chain.
+  ///
+  /// If more than one activator accepts the key event, then all of the
+  /// callbacks associated with activators that accept the key event are
+  /// invoked.
+  ///
+  /// Some examples of [ShortcutActivator] subclasses that can be used to define
+  /// the key combinations here are [SingleActivator], [CharacterActivator], and
+  /// [LogicalKeySet].
+  final Map<ShortcutActivator, VoidCallback> bindings;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    return Focus(
+      canRequestFocus: false,
+      skipTraversal: true,
+      onKey: (FocusNode node, RawKeyEvent event) {
+        KeyEventResult result = KeyEventResult.ignored;
+        // Activates all key bindings that match, returns "handled" if any handle it.
+        for (final ShortcutActivator activator in bindings.keys) {
+          if (activator.accepts(event, RawKeyboard.instance)) {
+            bindings[activator]!.call();
```

Comment:
Maybe it would be helpful to wrap this call with some error handling that produces an error which clarifies which key triggered the activator and the activator itself.

Reply:
What error am I handling?  Do you mean the `!` throwing because it found a `null`?  Because I don't think that can happen: The map has a non-nullable `VoidCallback`, and the `activator` is guaranteed to be in the map (because it's looping over `bindings.keys`).

Reply:
If the non-null function returned by `bindings[activator]` throws, you could help the developer out by reporting  which key triggered the activator and the activator itself.

Reply:
Ahh, OK, makes sense.

Reply:
Actually, in trying to do this, it seems like I'll either need to just print something (in debug mode only) and rethrow, or I'll have to throw a different exception from the one thrown.

Changing the exception type seems bad, since I don't know what could be thrown or who is listening for that exception (I'm thinking about things like AssertionError that aren't really meant to be caught), and rethrowing it isn't something we do anywhere else in the codebase, and might bypass all of the fancy code we have to avoid duplicate stack traces in the output.

Which path do you think would be best?

Reply:
If you factored out a function with a helpful name, the debug-mode stack trace might be more helpful:
```
void _applyKeyBinding(ShortcutActivator activator, RawKeyEvent event) {
  if (activator.accepts(event, RawKeyboard.instance)) {
    bindings[activator]!.call();
  }
}
```

Reply:
I checked out this branch and tried throwing an error inside of one of these VoidCallbacks.  The error is logged to the console and it does properly pinpoint the line in my main.js where I'm throwing the error.  For the developer, it will be easy to track down which keyboard shortcut caused the error, though I guess it could be nice to automatically have that in the logs without having to lookup the line of code.

Reply:
Cool, thanks. I think adding the helper function is a great way to go: we don't need to handle the exceptions at all, it just places the arguments into the stack so they can see what's being processed when an exception is thrown.